docs: update readme for usage of @GeneratedValue and select for update (#1785)
GoogleCloudPlatform/google-cloud-spanner-hibernate: README.adoc

Leaving out this annotation causes a `BYTES` column to be used.

==== Use @GeneratedValue for ID Generation
__NOTE__: Read to the end of this section to see the recommended way to set up `@GeneratedValue`.

❌ Not recommended

> `GenerationType.AUTO`, `GenerationType.TABLE`, `GenerationType.SEQUENCE` - Not supported by the Spanner dialect.

Reviewer: nit:

```suggestion
`GenerationType.AUTO`, `GenerationType.TABLE`, `GenerationType.SEQUENCE`: These are not supported by the Spanner dialect.
```

Note for the longer term (so not for this PR):
1. We should support AUTO. That should lead to the 'least-bad' auto-generation strategy that we can support, which I believe with our current feature set would be IDENTITY.
2. We should support TABLE if possible. I believe that the current limitation is that Hibernate generates an initialization command that tries to insert a row without specifying the column name. We should try to either override that in our dialect, or change this in Hibernate itself.

`GenerationType.IDENTITY`
-   See https://cloud.google.com/spanner/docs/primary-key-default-value#identity-columns[Identity columns].

> IDs are generated at insert time, which prevents Hibernate from batching inserts efficiently.

Reviewer: ```suggestion
It is not recommended to use the `IDENTITY` generation strategy, as it prevents Hibernate from using JDBC batching for inserts. See https://docs.jboss.org/hibernate/orm/current/userguide/html_single/Hibernate_User_Guide.html#batch-session-batch for more information.
```

`@GeneratedValue` +  `@SequenceGenerator` with `allocationSize = 1`

> Uses https://cloud.google.com/spanner/docs/primary-key-default-value#bit-reversed-sequence[Bit-reversed sequences] by spanner. Batching is inhibited because each insert requires a round‑trip to database

Reviewer: ```suggestion
This strategy uses a https://cloud.google.com/spanner/docs/primary-key-default-value#bit-reversed-sequence[bit-reversed sequence] with an allocation size of 1. This requires Hibernate to fetch a new
sequence value for each insert, causing each insert to cause two round-trips to Spanner.
```

`@GeneratedValue` +  `@SequenceGenerator` with `allocationSize = 1`
Uses https://cloud.google.com/spanner/docs/primary-key-default-value#bit-reversed-sequence[Bit-reversed sequences] by spanner. Batching is inhibited because each insert requires a round‑trip to database
----
@Entity
public class Employee {

  // Generates a bit-reversed sequence with an increment_size=1.
  // This is not recommended!
  @Id
  @GeneratedValue   
  @GeneratedValue(generator = "employee_generator")
  @SequenceGenerator(
      name = "employee_generator",
      sequenceName = "employee_seq",
      allocationSize = 1)
  public Long id;
}
----

Bit-reversed sequences do not support an increment size larger than 1. This means that entities
that use this style of identifiers by default require a round-trip to the database for each entity
that is inserted. The `PooledBitReversedSequenceStyleGenerator` provided in this repository fixes
this problem by using the configured `increment_size` to generate a query that fetches multiple
identifier values from the sequence in one query, instead of setting an `increment_size` on the

> sequence in the database.

Reviewer: Why do we want to remove this?

=== Locking

> Cloud Spanner supports exclusive row locks using https://cloud.google.com/spanner/docs/use-select-for-update[`SELECT ... FOR UPDATE`]

Reviewer: ```suggestion
Cloud Spanner supports explicit row locks using https://cloud.google.com/spanner/docs/use-select-for-update[`SELECT ... FOR UPDATE`]
```

These locks are not 'exclusive', as they do allow reads to continue to happen on these rows.
